crates/environ/src/fact/trampoline.rs

```diff
@@ -80,50 +84,192 @@ struct Compiler<'a, 'b> {
+        let lower_sig = module.types.signature(&adapter.lower, Context::Lower);
+        let lift_sig = module.types.signature(&adapter.lift, Context::Lift);
+        let ty = module
+            .core_types
+            .function(&lower_sig.params, &lower_sig.results);
+        let result = module
+            .funcs
+            .push(Function::new(Some(adapter.name.clone()), ty));
+
+        // If this type signature contains any borrowed resources then invocations
+        // of enter/exit call for resource-related metadata tracking must be used.
+        // It shouldn't matter whether the lower/lift signature is used here as both
+        // should return the same answer.
+        let emit_resource_call = module.types.contains_borrow_resource(&adapter.lower);
+        assert_eq!(
+            emit_resource_call,
+            module.types.contains_borrow_resource(&adapter.lift)
+        );
+
+        (
+            Compiler::new(module, result, lower_sig.params.len() as u32),
+            lower_sig,
+            lift_sig,
+        )
+    }
+
+    let async_start_adapter = |module: &mut Module, param_globals| {
+        let sig = module.types.async_start_signature(&adapter.lift);
+        let ty = module.core_types.function(&sig.params, &sig.results);
+        let result = module.funcs.push(Function::new(
+            Some(format!("[async-start]{}", adapter.name)),
+            ty,
+        ));
+
+        Compiler::new(module, result, sig.params.len() as u32).compile_async_start_adapter(
+            adapter,
+            &sig,
+            param_globals,
+        );
+
+        result
+    };
+
+    let async_return_adapter = |module: &mut Module, result_globals| {
+        let sig = module.types.async_return_signature(&adapter.lift);
+        let ty = module.core_types.function(&sig.params, &sig.results);
+        let result = module.funcs.push(Function::new(
+            Some(format!("[async-return]{}", adapter.name)),
+            ty,
+        ));
+
+        Compiler::new(module, result, sig.params.len() as u32).compile_async_return_adapter(
+            adapter,
+            &sig,
+            result_globals,
+        );
+
+        result
+    };
+
     match (adapter.lower.options.async_, adapter.lift.options.async_) {
-        (false, false) => {}
+        (false, false) => {
+            // We can adapt sync->sync case with only minimal use of intrinsics,
+            // e.g. resource enter and exit calls as needed.
+            let (compiler, lower_sig, lift_sig) = compiler(module, adapter);
+            compiler.compile_sync_to_sync_adapter(adapter, &lower_sig, &lift_sig)
+        }
         (true, true) => {
-            todo!()
+            // In the async->async case, we must compile a couple of helper functions:
+            //
+            // - `async-start`: copies the parameters from the caller to the callee
+            // - `async-return`: copies the result from the callee to the caller
```

Comment:
One question I'm left with after reading these docs is what's done with async-{start,return}, e.g. who's calling them. Definitely feel free to say "go read the CM docs Alex" for this, but it's at least to me still a big missing piece of the puzzle in understanding these trampolines.